File: Priority_queue.java
package learnjava;

import java.util.Vector;

class Heap{
    private Vector<Integer> heap = new Vector<Integer>();

    public void push(int elem){
        heap.add(elem);
        heapify_up(heap.size()-1);
    }

    public void pop(){
        heap.remove(0);
        heapify_down(0);
    }

    public int top(){ return heap.get(0); }
    public int size(){ return heap.size(); }
    public boolean empty(){ return heap.size() == 0; }

    public void heapify_up(int index){
        while(index > 0){
            int parent = (int)(1.0*(index-1)/2);
            if (heap.get(parent) > heap.get(index)){
                int temp = heap.get(parent);
                heap.set(parent, heap.get(index));
                heap.set(index, temp);
                index = parent;
            }
            else break;
        }
    }

    public void heapify_down(int index){
        int left = 2*index + 1;
        int right = 2*index + 2;
        int min = index;

        if (left < heap.size() && heap.get(left) < heap.get(min))
            min = left;
        if (right < heap.size() && heap.get(right) < heap.get(min))
            min = right;

        if (index != min){
            int temp = heap.get(index);
            heap.set(index, heap.get(min));
            heap.set(min, temp);
            heapify_down(min);
        }
    }
}

public class Priority_queue {
    private Heap heap_;

    public Priority_queue(){
        heap_ = new Heap();
    }

    public int size(){ return heap_.size(); }
    public boolean empty(){ return heap_.empty(); }
    public void push(int elem){ heap_.push(elem); }
    public void pop(){ heap_.pop(); }
    public int top(){ return heap_.top(); }

    public static void main(String[] args) {
        Priority_queue queue = new Priority_queue();

        queue.push(7);
        queue.push(2);
        queue.push(3);
        queue.push(1);
        queue.push(4);
        queue.push(1);

        while(!queue.empty()){
            System.out.println(queue.top());
            queue.pop();
        }
    }
}

File: priority_queue.cpp
#include <vector>	
#include "compare.cpp"

#include <algorithm>
	
#ifndef priority_queue_cpp
#define priority_queue_cpp

using namespace std;
	
template <class T, class Compare>
class PriorityQueue
{
private:
	vector<T> heap;
	Compare compare;
		
public:
	PriorityQueue(){ }
		
	void push(T elem){
		heap.push_back(elem);
		heapify_up(heap.size()-1);
	}
		
	void pop(){
		if (!heap.empty()){
			swap(heap[0], heap[heap.size()-1]);
			heap.pop_back();
			heapify_down(0);
			
		} else return;
	}
		
	T top(){ return heap.front(); }
	int size(){ return heap.size(); }	
	bool empty(){ return heap.empty(); }
	
private:	
	void heapify_down(int index);
	void heapify_up(int index);
};
	
template<class T, class Compare>
void PriorityQueue<T, Compare>::heapify_down(int index) // update dong khi xoa phan tu dau tien
{
	int left = 2*index+1; 
	int right = 2*index+2; 
	int tmp = index;
	
	if (left < heap.size() && compare(heap[left], heap[tmp]))
		tmp = left;
		
	if (right < heap.size() && compare(heap[right], heap[tmp]))
		tmp = right;
		
	if (tmp != index){
		swap(heap[index], heap[tmp]);
		heapify_down(tmp);
	}
}
	
template<class T, class Compare>
void PriorityQueue<T, Compare>::heapify_up(int index) // update dong khi them mot phan tu va heap
{
	while(index > 0){
		int parent = 1.0*(index-1)/2;
		if (compare(heap[index], heap[parent])){
			swap(heap[parent], heap[index]);
			index = parent;
		}
		else break;
	}
}

#endif